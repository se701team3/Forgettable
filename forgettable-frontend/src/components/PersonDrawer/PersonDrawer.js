import React from 'react';
import classes from './PersonDrawer.module.css';
import {Avatar, Drawer} from '@mui/material';
import PropTypes from 'prop-types';
import {
  calculateAge,
  getDateLastMetString,
  getFirstMetTimeString,
} from '../../functions/dateFormatter';
import {capitalise} from '../../functions/stringFormatter';
import convertSocialMediaNamesToIcons,
{convertSocialMediaToIcon} from '../../functions/socialMediaIconConverter';
import {IconButton} from '@mui/material';
import {getBirthdayString} from '../../functions/dateFormatter';
import CustomButton from '../CustomButton/CustomButton';
import classNames from 'classnames';
import UnknownDetail from '../UnknownDetail/UnknownDetail';

const PersonDrawer = (props) => {
  return (
    <Drawer
      sx={{
        'width': props.staticDrawer ? '375px' : '460px',
        'flexShrink': 0,
        '& .MuiDrawer-paper': {
          width: props.staticDrawer ? '375px' : '460px',
          boxSizing: 'border-box',
          marginLeft: '128px',
        },
      }}
      variant="persistent"
      anchor={props.staticDrawer ? 'left' : 'right'}
      open={props.open}
      data-testid="drawer-element"
    >
      <div className={classes.PersonDrawer}>
        <div className={classes.ContentContainer}>
          <Avatar
            alt={props.name}
            // this style is written inline because MUI does not support className
            style={{
              height: '200px',
              width: '200px',
              backgroundColor:
                getComputedStyle(document.body)
                    .getPropertyValue('--prmry'),
              fontSize:
                getComputedStyle(document.body)
                    .getPropertyValue('--text-xxlarge')
              ,
            }}
            src={props.img}
          />
          <div className={classes.InfoHeader}>
            <h1 data-testid="name-element">
              {props.name}
            </h1>
            <h2 data-testid="first-met-element">
              First met {getFirstMetTimeString(props.firstMet)}
            </h2>
          </div>
          <div
            className={props.staticDrawer ?
          classes.InfoContent :
          classNames(classes.InfoContent, classes.InfoContentPadding)}
          >
            <p data-testid="age-element">
              {'Age: '}
              {props.birthday?
              <span className={classes.KnownText}>
                {calculateAge(props.birthday)}
              </span> :
              <UnknownDetail/>}
            </p>
            <p data-testid="gender-element">
              {'Gender: '}
              {props.gender ?
              <span className={classes.KnownText}>
                {capitalise(props.gender)}
              </span> :
              <UnknownDetail/>}
            </p>
            <p data-testid="birthday-element">
              {'Birthday: '}
              {props.birthday ?
              <span className={classes.KnownText}>
                {getBirthdayString(props.birthday)}
              </span> :
              <UnknownDetail/>}
            </p>
            <p data-testid="organisation-element">
              {'Organisation: '}
              {props.organisation ?
              <span className={classes.KnownText}>
                {capitalise(props.organisation)}
              </span> :
              <UnknownDetail/>}
            </p>
            <p data-testid="location-element">
              {' Location: '}
              {props.location ?
              <span className={classes.KnownText}>
                {capitalise(props.location)}
              </span> :
              <UnknownDetail/>}
            </p>
            <p data-testid="date-first-met-element">
              {'First met: '}
              {props.firstMet ?
                <span className={classes.KnownText}>
                  {getDateLastMetString(props.firstMet)}
                </span> :
                 <UnknownDetail/>}
            </p>
            <p data-testid="interests-element">
              {'Interests: '}
              {props.interests ?
               <span className={classes.KnownText}>
                 {props.interests.join(', ')}
               </span> :
               <UnknownDetail/>}
            </p>
            <p data-testid="social-medias-element">
              {'Social media: '}
              {props.socialMedias?.length ?
               <span className={classes.SocialMediaContainer}>
                 {props.socialMedias.map((socialMedia) => {
                   const icon = convertSocialMediaToIcon(socialMedia.name);

                   return (icon ?
                    <IconButton
                      key={socialMedia.name}
                      data-testid={`social-media-element-${socialMedia.name}`}
                      onClick={() => window.open(socialMedia.link, '_blank')}
                    >
                      <img src={icon} alt={socialMedia.name} />
                    </IconButton> :
                    null
                   );
                 })}
               </span> :<UnknownDetail/>}
            </p>
            {props.staticDrawer &&
            <CustomButton
              btnText="Edit"
              className={classes.EditButton}
              textStyle={classes.ButtonText}
              onClick={props.onEdit}
            />}
          </div>
        </div>
      </div>
    </Drawer>
  );
};

PersonDrawer.propTypes = {
  open: PropTypes.bool.isRequired,
  id: PropTypes.string,
  name: PropTypes.string.isRequired,
  img: PropTypes.string,
  socialMedias: PropTypes.arrayOf(Object), // See below for an example
  // [{name: 'facebook', link: 'https://www.google.com/'}, {name: 'instagram', link: 'https://www.google.com/'}]
  firstMet: PropTypes.instanceOf(Date),
  location: PropTypes.string,
  interests: PropTypes.arrayOf(PropTypes.string),
  staticDrawer: PropTypes.bool,
  onEdit: PropTypes.bool,
};

export default PersonDrawer;
