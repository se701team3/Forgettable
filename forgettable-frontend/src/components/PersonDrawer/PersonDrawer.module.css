@import '../../colours.css';

.PersonDrawer {
    height: 100%;
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    box-sizing: border-box;
    padding: 55px;
}

.ContentContainer {
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
}

.InfoHeader {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 30px 0;
}

.InfoHeader > h1 {
    font-size: var(--text-drawer-title);
    font-weight: 600;
    margin: 5px 0;
    text-align: center;
}

.InfoHeader > h2 {
    font-size: var(--text-medium);
    font-weight: 400;
    color: var(--txt2);
    margin: 0;
    text-align: center;
}

.InfoContent {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    box-sizing: border-box;
    padding: 0;
    align-self: flex-start;
}

.InfoContentPadding {
    padding: 0 50px;
}

.InfoContent > p {
    font-size: var(--text-medium);
    font-weight: 400;
    margin: 15px 0;
    color: var(--txt2);
}

.KnownText {
    font-weight: 500;
    color: var(--txt3);
}

.UnknownText {
    color: var(--bord);
}

.SocialMediaContainer {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: flex-start;
}

.Icon {
    height: 36px;
    width: 36px;
}

.EditButton {
    width: 80px;
    margin-top: 25px;
}

.ButtonText {
    font-size: var(--text-medium);
}