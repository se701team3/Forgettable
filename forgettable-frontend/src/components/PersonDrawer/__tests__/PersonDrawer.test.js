import TestRenderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import React from 'react';
import {render, fireEvent, cleanup} from '@testing-library/react';
import {
  within,
  getByText,
  getByTestId,
} from '@testing-library/dom';
import PersonDrawer from '../PersonDrawer';

const PERSON_NAME = 'Mercury Lin';

it('renders PersonDrawer UI with correct hierarchy', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<PersonDrawer
    open={true}
    name="Mercury Lin"
    img="https://user-images.githubusercontent.com/62003343/157863205-d299d1aa-deeb-4009-bcc8-2238227f5703.gif"
    firstMet={new Date().now}
    location="London"
    gender="female"
    organisation="Team3"
    interests={['food, sleep']}
    socialMedias={[
      {
        name: 'facebook',
        link: 'https://google.com/',
      },
      {
        name: 'instagram',
        link: 'https://google.com/',
      },
    ]
    }
  />);
  const result = renderer.getRenderOutput();

  expect(result).toMatchSnapshot();
});

afterEach(cleanup);

it('Name prop passed in is displayed correctly', () => {
  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
      />);

  const {getByText} = within(getByTestId('name-element'));

  expect(getByText(PERSON_NAME)).toBeInTheDocument();
});

it('displays the correct format for firstMet for an existing date', () => {
  const twoYearsAgo = new Date(Date.now() - (365 * 24 * 60 * 60 * 1000 * 2));

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        firstMet={twoYearsAgo}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met 2 years ago'),
  ).toBeInTheDocument();
});

it('displays the firstMet string when there is no date', () => {
  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
      />);

  const {getByText} = within(getByTestId('first-met-element'));

  expect(
      getByText('First met once upon a time'),
  ).toBeInTheDocument();
});

it('displays the correct age number given a timestamp', () => {
  const birthday = new Date('2000-01-01');

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        birthday={birthday}
      />);

  const {getByText} = within(getByTestId('birthday-element'));

  expect(
      getByText('01 Jan 2000'),
  ).toBeInTheDocument();
});

it('displays the correct organisation', () => {
  const org = 'Team3';

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        organisation={org}
      />);

  const {getByText} = within(getByTestId('organisation-element'));

  expect(
      getByText(org),
  ).toBeInTheDocument();
});

it('displays the correct organisation', () => {
  const location = 'London';

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        location={location}
      />);

  const {getByText} = within(getByTestId('location-element'));

  expect(
      getByText(location),
  ).toBeInTheDocument();
});


it('displays the correct date first met in detailed fields', () => {
  const date = new Date('01/01/2000');

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        firstMet={date}
      />);

  const {getByText} = within(getByTestId('date-first-met-element'));

  expect(
      getByText('01/01/2000'),
  ).toBeInTheDocument();
});


it('displays the correct list of interests in correct format', () => {
  const interests = ['food', 'sleep'];

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        interests={interests}
      />);

  const {getByText} = within(getByTestId('interests-element'));

  expect(
      getByText('food, sleep'),
  ).toBeInTheDocument();
});

it('displays the social media items correctly', () => {
  const socialMedias =
    [
      {
        name: 'facebook',
        link: 'https://github.com/',
      },
      {
        name: 'instagram',
        link: 'https://github.com/',
      },
    ];

  const {getByTestId} = render(
      <PersonDrawer
        open={true}
        name={PERSON_NAME}
        socialMedias={socialMedias}
      />);

  // eslint-disable-next-line guard-for-in
  for (const i in socialMedias) {
    expect(
        getByTestId(
            `social-media-element-${socialMedias[i].name}`,
        ),
    ).toBeTruthy();
  }
});

