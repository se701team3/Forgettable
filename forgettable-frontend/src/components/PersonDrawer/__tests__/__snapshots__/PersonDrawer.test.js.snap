// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PersonDrawer UI with correct hierarchy 1`] = `
<ForwardRef(Drawer)
  anchor="right"
  data-testid="drawer-element"
  open={true}
  sx={
    Object {
      "& .MuiDrawer-paper": Object {
        "boxSizing": "border-box",
        "marginLeft": "128px",
        "width": "460px",
      },
      "flexShrink": 0,
      "width": "460px",
    }
  }
  variant="persistent"
>
  <div
    className="PersonDrawer"
  >
    <div
      className="ContentContainer"
    >
      <ForwardRef(Avatar)
        alt="Mercury Lin"
        src="https://user-images.githubusercontent.com/62003343/157863205-d299d1aa-deeb-4009-bcc8-2238227f5703.gif"
        style={
          Object {
            "backgroundColor": "",
            "fontSize": "",
            "height": "200px",
            "width": "200px",
          }
        }
      />
      <div
        className="InfoHeader"
      >
        <h1
          data-testid="name-element"
        >
          Mercury Lin
        </h1>
        <h2
          data-testid="first-met-element"
        >
          First met 
          once upon a time
        </h2>
      </div>
      <div
        className="InfoContent InfoContentPadding"
      >
        <p
          data-testid="age-element"
        >
          Age: 
          <span
            className="UnknownText"
          >
            Unknown
          </span>
        </p>
        <p
          data-testid="gender-element"
        >
          Gender: 
          <span
            className="KnownText"
          >
            Female
          </span>
        </p>
        <p
          data-testid="birthday-element"
        >
          Birthday: 
          <span
            className="UnknownText"
          >
            Unknown
          </span>
        </p>
        <p
          data-testid="organisation-element"
        >
          Organisation: 
          <span
            className="KnownText"
          >
            Team3
          </span>
        </p>
        <p
          data-testid="location-element"
        >
           Location: 
          <span
            className="KnownText"
          >
            London
          </span>
        </p>
        <p
          data-testid="date-first-met-element"
        >
          First met: 
          <span
            className="UnknownText"
          >
            Unknown
          </span>
        </p>
        <p
          data-testid="interests-element"
        >
          Interests: 
          <span
            className="KnownText"
          >
            food, sleep
          </span>
        </p>
        <p
          data-testid="social-medias-element"
        >
          Social media: 
          <span
            className="SocialMediaContainer"
          >
            <ForwardRef(IconButton)
              data-testid="social-media-element-facebook"
              onClick={[Function]}
            >
              <img
                alt="facebook"
                src="facebook.svg"
              />
            </ForwardRef(IconButton)>
            <ForwardRef(IconButton)
              data-testid="social-media-element-instagram"
              onClick={[Function]}
            >
              <img
                alt="instagram"
                src="instagram.svg"
              />
            </ForwardRef(IconButton)>
          </span>
        </p>
      </div>
    </div>
  </div>
</ForwardRef(Drawer)>
`;
